worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,142 @@
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "admin"
+)
+
+func (w *Worker) registerHandlers(handle func(pattern string, handler http.Handler)) {
+	// TODO: does this API structure make sense?
+	handle("/metrics-users/add", http.HandlerFunc(w.handleAddMetricsUser))
+	handle("/metrics-users/remove", http.HandlerFunc(w.handleRemoveMetricsUser))
+}
+
+func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	queryParams := req.URL.Query()
+
+	if !queryParams.Has("username") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "username" query parameter`)
+		return
+	}
+	username := queryParams.Get("username")
+
+	if !queryParams.Has("password") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "password" query parameter`)
+		return
+	}
+	password := queryParams.Get("password")
+
+	err := validateMetricsUsername(username)
+	if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	if password == "" {
+		w.writeResponse(resp, http.StatusBadRequest, "empty password")
+		return
+	}
+
+	user, err := w.state.AddUser(username, username, password, userCreator)
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "failed to create user %q: %v", username, err)
+		return
+	}
+	userTag := user.UserTag()
+	// We should clean up the user if the later grant access fails.
+	cleanup := func() {
+		err := w.state.RemoveUser(userTag)
+		if err != nil {
+			// Best we can do here is log an error.
+			w.logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+				username, err)
+		}
+	}
+
+	model, err := w.state.Model()
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "retrieving current model: %v", err)
+		cleanup()
+		return
+	}
+
+	_, err = model.AddUser(state.UserAccessSpec{
+		User:      userTag,
+		CreatedBy: names.NewUserTag(userCreator),
+		Access:    permission.ReadAccess,
+	})
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "adding user %q to model %q: %v",
+			username, bootstrap.ControllerModelName, err)
+		return
+	}
+
+	w.writeResponse(resp, http.StatusOK, "successfully added metrics user %q", username)
+}
+
+func (w *Worker) handleRemoveMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	queryParams := req.URL.Query()
+
+	if !queryParams.Has("username") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "username" query parameter`)
+		return
+	}
+	username := queryParams.Get("username")
+
+	err := validateMetricsUsername(username)
+	if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	userTag := names.NewUserTag(username)
+	err = w.state.RemoveUser(userTag)
+	if errors.Is(err, errors.UserNotFound) {
+		w.writeResponse(resp, http.StatusNotFound, err.Error())
```

Comment:
This should be idempotent. Return OK.

Reply:
point of contention. I think Ian would disagree with you :P

Reply:
The problem is that you can then use an enumeration attack to find user names. As you're now saying, any user that doesn't exist will tell you that they're not found and you'll succeed using a brute force attack. Sending OK doesn't completely mitigate this, but it does make it harder, as you now need to use timing attacks to do so.

This should be the basic setup for all security-based endpoints. Do not expose your database via your API.

Reply:
OK, you've convinced me based on that

Reply:
This is less of a concern if we have the paths authenticated. Otherwise you could do the same attack with `GET` requests.

When I proposed the control socket as a solution, it was always my intent that we would:
- Password-protect the paths (no statement on cardinality yet)
- Have the controller generate a secret for use by the charm to authenticate itself for socket usage.
- Have the charm acquire and use the secret.

Reply:
We must remember here that this API is provided by the controller on a unix socket protected by fs permissions for the controller charm who is supposed to drive the controller.

If you want to enforce the connecting socket is indeed the same user as the controller, you can use ucred socket option as seen here https://github.com/canonical/pebble/blob/master/internals/daemon/ucrednet.go
But this is only needed for multi user access (in the case of the snap daemon), as the socket is wide open to all.

Simply creating the socket with no group or other access and ensuring the socket is owned by the jujud processes' user is enough here.

With regards to using the controller to generate a secret for the charm to talk to the controller, this introduces more chicken and egg problems in the future, which is part of the work we are trying to avoid here. 
If we want the controller charm to eventually own the jujud instance, to manage it, we should use very basic yet secure mechanisms that are a common pattern in linux and in canonical (see pebble, snapd and lxd).